packages/dev/src/lib/env.ts

```diff
@@ -0,0 +1,427 @@
+import type { NetlifyAPI } from '@netlify/api'
+import type { EnvironmentVariables } from '@netlify/types'
+
+export type EnvironmentVariableSource = 'account' | 'addons' | 'configFile' | 'general' | 'internal' | 'ui'
+
+/**
+ * Supported values for the user-provided env `context` option.
+ * These all match possible `context` values returned by the Envelope API.
+ * Note that a user may also specify a branch name with the special `branch:my-branch-name` format.
+ */
+export const SUPPORTED_CONTEXTS = ['all', 'production', 'deploy-preview', 'branch-deploy', 'dev'] as const
+
+/**
+ * Additional aliases for the user-provided env `context` option.
+ */
+const SUPPORTED_CONTEXT_ALIASES = {
+  dp: 'deploy-preview',
+  prod: 'production',
+}
+/**
+ * Supported values for the user-provided env `scope` option.
+ * These exactly match possible `scope` values returned by the Envelope API.
+ * Note that `any` is also supported.
+ */
+export const ALL_ENVELOPE_SCOPES = ['builds', 'functions', 'runtime', 'post_processing'] as const
+
+export type EnvironmentVariable = {
+  sources: EnvironmentVariableSource[]
+  value: string
+  scopes?: (typeof ALL_ENVELOPE_SCOPES)[]
+}
+
+type InjectedEnvironmentVariable = {
+  isInternal: boolean
+  originalValue: string | undefined
+  overriddenSources: string[]
+  usedSource: string
+}
+
+interface InjectEnvironmentVariablesOptions {
+  accountSlug?: string
+  baseVariables: Record<string, EnvironmentVariable>
+  envAPI: EnvironmentVariables
+  netlifyAPI?: NetlifyAPI
+  siteID?: string
+}
+
+export const injectEnvVariables = async ({
```

Comment:
This file was also copied from the CLI, but there are some changes in this method:

- It uses the `EnvironmentVariables` abstraction as opposed to interacting with `process` directly, which makes the process of restoring the original value of the environment variables automatic
- It no longer logs the variables that were added, and instead simply returns an object with all the information needed for any consumers to handle the logging

Reply:
Pretty much the two remaining refactors I wanted to do when I was refactoring this file a few weeks ago but punted on for time 💯